Strange `-Wcast-align` warning from `clang-3.9`

I saw the following strange warning (as an error) from `clang-3.9`.  ( Here's [actual GH-Actions log](https://github.com/t-mat/xxHash/runs/3345251571#step:7:23) )

```
clang-3.9 -Werror -Wall -Wextra -Wconversion -Wcast-qual -Wcast-align -Wshadow \
          -Wstrict-aliasing=1 -Wswitch-enum -Wdeclaration-after-statement \
          -Wstrict-prototypes -Wundef -Wpointer-arith -Wformat-security -Wvla \
          -Wformat=2 -Winit-self -Wfloat-equal -Wwrite-strings -Wredundant-decls \
          -Wstrict-overflow=2 -Wno-sign-conversion -DXXHSUM_DISPATCH=1 -c -o xxhash.o xxhash.c

In file included from xxhash.c:43:
./xxhash.h:3868:56: error:
    cast from 'const xxh_u8 *' (aka 'const unsigned char *') to 'const float *'
    increases required alignment from 1 to 4 [-Werror,-Wcast-align]

    XXH_ALIGN(64)        const float* const src  = (float const*) XXH3_kSecret;
                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I'd like to know there's good solution for this issue.

I made the following minimum reproduction code

```c
#include <stdio.h>
#include <stdalign.h>
#include <stdint.h>

#define XXH_ALIGN(n) __attribute__ ((aligned(n)))
typedef uint8_t xxh_u8;
XXH_ALIGN(64) static const xxh_u8 XXH3_kSecret[192] = { 0 };

int main(int argc, char** argv) {
    XXH_ALIGN(64) const float* const src = (const float* const) XXH3_kSecret;
    printf("alignof(XXH3_kSecret)=%zd\n", alignof(XXH3_kSecret));
    printf("alignof(src)=%zd\n", alignof(src));
}
```

----
[clang-3.9.1](https://wandbox.org/permlink/aZgUDs8HrwZdzsKz) generates the following warning which I saw in GH-Actions.

```
clang prog.c -Wall -Wextra -pedantic "-Wcast-align" "-Wno-gnu-alignof-expression" "-Wno-unused-parameter"

prog.c:10:41: warning: cast from 'const xxh_u8 *' (aka 'const unsigned char *') to 'const float *const' increases required alignment from 1 to 4 [-Wcast-align]
        XXH_ALIGN(64) const float* const src = (const float* const) XXH3_kSecret;
                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.

alignof(XXH3_kSecret)=64
alignof(src)=64
```

Note that `alignof()` shows expected and correct value `64`.

----
[clang-4.0.0](https://wandbox.org/permlink/jBSkBNW3tTow9f2J) and later versions compile it without warning though.

```
clang prog.c -Wall -Wextra -std=c89 "-Wcast-align" "-Wno-gnu-alignof-expression" "-Wno-unused-parameter"

alignof(XXH3_kSecret)=64
alignof(src)=64
```

> Why is the src of type const float* ?

Actual prototype of `_mm_load_ps()` is `__m128 _mm_load_ps(const float *p)`.  It requires `const float*`.

> Could it be done differently ?

Perhaps, we can employ `union` trick which is implemented in `XXH3_initCustomSecret_avx512()`.

https://github.com/Cyan4973/xxHash/blob/4a20afceb625f62278e11f630156475aee40b055/xxhash.h#L3668-L3673

```c
// clang-3.9 prog.c -Wall -Wextra -std=gnu11 "-Wcast-align" "-Wno-gnu-alignof-expression" "-Wno-unused-parameter"
#include <stdio.h>
#include <stdalign.h>
#include <stdint.h>

#define XXH_ALIGN(n) __attribute__ ((aligned(n)))
typedef uint8_t xxh_u8;
XXH_ALIGN(64) static const xxh_u8 XXH3_kSecret[192] = { 0 };

int main(int argc, char** argv) {
    union {
        XXH_ALIGN(64) const float* cfp;
        XXH_ALIGN(64) const uint8_t* cu8p;
    } temp;
    temp.cu8p = XXH3_kSecret;

    XXH_ALIGN(64) const float* src  = temp.cfp;                    // no warning
    XXH_ALIGN(64) const float* src2 = (const float*) XXH3_kSecret; // warning [-Wcast-align]

    printf("alignof(XXH3_kSecret)=%zd\n", alignof(XXH3_kSecret));
    printf("alignof(src)         =%zd\n", alignof(src));
    printf("alignof(src2)        =%zd\n", alignof(src2));
}
```

result: https://wandbox.org/permlink/LBqz1yP7o600nbjJ

```
prog.c:17:39: warning: cast from 'const xxh_u8 *' (aka 'const unsigned char *') to 'const float *' increases required alignment from 1 to 4 [-Wcast-align]
    XXH_ALIGN(64) const float* src2 = (const float*) XXH3_kSecret; // warning [-Wcast-align]
                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
alignof(XXH3_kSecret)=64
alignof(src)         =64
alignof(src2)        =64
```

We need to benchmark/investigate there's no penalty in this change though.

--- a/xxhash.h
+++ b/xxhash.h
@@ -3865,19 +3865,21 @@ XXH_FORCE_INLINE XXH_TARGET_SSE2 void XXH3_initCustomSecret_sse2(void* XXH_RESTR
 #       endif
         int i;
 
-        const __m128i* const src16 = (const __m128i *)XXH3_kSecret;
-        __m128i* dest = (__m128i*) customSecret;
+        const void* const src16 = XXH3_kSecret;
+        __m128i* dst16 = (__m128i*) customSecret;
 #       if defined(__GNUC__) || defined(__clang__)
         /*
          * On GCC & Clang, marking 'dest' as modified will cause the compiler:
          *   - do not extract the secret from sse registers in the internal loop
          *   - use less common registers, and avoid pushing these reg into stack
          */
-        XXH_COMPILER_GUARD(dest);
+        XXH_COMPILER_GUARD(dst16);
 #       endif
+        XXH_ASSERT(((size_t)src16 & 15) == 0); /* control alignment */
+        XXH_ASSERT(((size_t)dst16 & 15) == 0);
 
         for (i=0; i < nbRounds; ++i) {
-            dest[i] = _mm_add_epi64(_mm_load_si128(src16+i), seed);
+            dst16[i] = _mm_add_epi64(_mm_load_si128((const __m128i *)src16+i), seed);
     }   }
 }
 

--- a/xxhash.h
+++ b/xxhash.h
@@ -3865,8 +3865,8 @@ XXH_FORCE_INLINE XXH_TARGET_SSE2 void XXH3_initCustomSecret_sse2(void* XXH_RESTR
 #       endif
         int i;
 
-        XXH_ALIGN(64)        const float* const src  = (float const*) XXH3_kSecret;
-        XXH_ALIGN(XXH_SEC_ALIGN) __m128i*       dest = (__m128i*) customSecret;
+        const __m128i* const src16 = (const __m128i *)XXH3_kSecret;
+        __m128i* dest = (__m128i*) customSecret;
 #       if defined(__GNUC__) || defined(__clang__)
         /*
          * On GCC & Clang, marking 'dest' as modified will cause the compiler:
@@ -3877,7 +3877,7 @@ XXH_FORCE_INLINE XXH_TARGET_SSE2 void XXH3_initCustomSecret_sse2(void* XXH_RESTR
 #       endif
 
         for (i=0; i < nbRounds; ++i) {
-            dest[i] = _mm_add_epi64(_mm_castps_si128(_mm_load_ps(src+i*4)), seed);
+            dest[i] = _mm_add_epi64(_mm_load_si128(src16+i), seed);
     }   }
 }
 

